crates/cli/src/subcommands/generate/csharp.rs

```diff
@@ -1,127 +1,533 @@
-
-fn scalar_or_string_name(b: &AlgebraicType) -> Option<&str> {
-    Some(match b {
-        AlgebraicType::Bool => "bool",
-        AlgebraicType::I8 => "sbyte",
-        AlgebraicType::U8 => "byte",
-        AlgebraicType::I16 => "short",
-        AlgebraicType::U16 => "ushort",
-        AlgebraicType::I32 => "int",
-        AlgebraicType::U32 => "uint",
-        AlgebraicType::I64 => "long",
-        AlgebraicType::U64 => "ulong",
-        AlgebraicType::I128 => "I128",
-        AlgebraicType::U128 => "U128",
-        AlgebraicType::I256 => "I256",
-        AlgebraicType::U256 => "U256",
-        AlgebraicType::String => "string",
-        AlgebraicType::F32 => "float",
-        AlgebraicType::F64 => "double",
-        AlgebraicType::Ref(_) | AlgebraicType::Sum(_) | AlgebraicType::Product(_) | AlgebraicType::Array(_) => {
-            return None
+impl Lang for Csharp<'_> {
+    fn table_filename(&self, _module: &ModuleDef, table: &TableDef) -> String {
+        format!("Tables/{}.g.cs", table.name.deref().to_case(Case::Pascal))
+    }
+
+    fn type_filename(&self, type_name: &spacetimedb_schema::def::ScopedTypeName) -> String {
+        format!("Types/{}.g.cs", collect_case(Case::Pascal, type_name.name_segments()))
+    }
+
+    fn reducer_filename(&self, reducer_name: &Identifier) -> String {
+        format!("Reducers/{}.g.cs", reducer_name.deref().to_case(Case::Pascal))
+    }
+
+    fn format_files(&self, generated_files: BTreeSet<PathBuf>) -> anyhow::Result<()> {
+        cmd!(
+            "dotnet",
+            "format",
+            // We can't guarantee that the output lives inside a valid project or solution,
+            // so to avoid crash we need to use the `dotnet whitespace --folder` mode instead
+            // of a full style-aware formatter. Still better than nothing though.
+            "whitespace",
+            "--folder",
+            // Our files are marked with <auto-generated /> and will be skipped without this option.
+            "--include-generated",
+            "--include"
+        )
+        .before_spawn(move |cmd| {
+            cmd.args(&generated_files);
+            Ok(())
+        })
+        .run()?;
+        Ok(())
+    }
+
+    fn generate_table(&self, module: &ModuleDef, table: &TableDef) -> String {
+        let mut output = CsharpAutogen::new(
+            self.namespace,
+            &[
+                "SpacetimeDB.BSATN",
+                "SpacetimeDB.ClientApi",
+                "System.Collections.Generic",
+                "System.Runtime.Serialization",
+            ],
+        );
+
+        writeln!(output, "public sealed partial class RemoteTables");
+        indented_block(&mut output, |output| {
+            let schema = TableSchema::from_module_def(module, table, (), 0.into())
+                .validated()
+                .expect("Failed to generate table due to validation errors");
+            let csharp_table_name = table.name.deref().to_case(Case::Pascal);
```

Comment:
This is actually way cleaner-looking than (*table.name).etc(), I just never thought to call it directly.

Reply:
Hah, I didn't come up with it - it was already used extensively in the codegen, so I just followed suit.